### PR TITLE
feat(aibom): add custom catalog for user-defined AI component detection

### PR DESCRIPTION
**Pull request by vineethsai7**
## Summary

- Adds a **custom catalog** system that lets users teach the analyzer about components the built-in DuckDB catalog does not cover, using three complementary mechanisms:
  1. **Configuration file** (`.aibom.yaml`/`.yml`/`.json`) -- register custom component symbols, base-class detection rules, exclude patterns, extended relationship argument hints, and custom relationship types.
  2. **Inline annotations** (`# aibom: concept=...`) -- tag classes and functions directly in source code for instant recognition.
  3. **Base class detection** -- automatically categorize any class that inherits from a user-specified base class.
- Supports **custom categories** (e.g. `guardrail`, `router`, `evaluator`), **exclude patterns** to suppress false positives, **metadata fields** on custom components, and **custom relationship types** (e.g. `ROUTES_TO`, `GUARDS`) with user-defined argument hints.
- Updates the CLI with a `--custom-catalog` option and auto-discovery of `.aibom.yaml` in source directories.

## Changes

| File | Description |
|------|-------------|
| `aibom/src/aibom/custom_catalog.py` | **New** -- Loader/validator for `.aibom.yaml`/`.json`, inline annotation parser |
| `aibom/src/aibom/structures.py` | Add `ClassDefObservation`, `FunctionAnnotationObservation`; extend `CodeAnalysisResult` |
| `aibom/src/aibom/cst_parser.py` | Add `leave_ClassDef`, update `leave_FunctionDef` for base class capture and `# aibom:` annotations |
| `aibom/src/aibom/catalog_db.py` | Add `add_custom_entries()`, `add_excludes()`, exclude filtering in queries |
| `aibom/src/aibom/categorizer.py` | Inline annotation pass, base class pass, custom categories, extended hints, custom relationship derivation |
| `aibom/src/aibom/cli.py` | `--custom-catalog` option, auto-discovery of `.aibom.yaml` |
| `aibom/pyproject.toml` | Add `pyyaml>=6.0` dependency |
| `README.md` | New **Custom Catalog** section with full examples, updated Features and Technical Details |
| `aibom/docs/TECHNICAL_OVERVIEW.md` | Updated architecture, execution flow, and module documentation |
| `aibom/examples/.aibom.yaml` | Annotated reference configuration file |
| `aibom/tests/test_custom_catalog.py` | **New** -- 28 tests covering config loading, inline annotations, base class detection, custom categories, excludes, metadata, CatalogDB integration |

## Test plan

- [x] All 28 new tests pass (`uv run pytest tests/test_custom_catalog.py -v`)
- [x] All 81 existing tests pass with zero regressions (`uv run pytest tests/ -v`)
- [ ] Manual: place `.aibom.yaml` in a project root and verify auto-discovery
- [ ] Manual: run with `--custom-catalog /path/to/.aibom.yaml` and verify custom components appear in output
- [ ] Manual: add `# aibom: concept=tool` inline annotation and verify detection
- [ ] Manual: verify exclude patterns suppress expected components

### CODE REVIEW

File: aibom/examples/.aibom.yaml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+
+# ─── 1. Custom Components ───────────────────────────────────────────────
+#
+# Register symbols (classes, functions, modules) that the built-in catalog
+# does not know about.
+#
+# Fields:
+#   id        (required)  Short class/function name or fully qualified name.
+#                          Short names use suffix matching, so "MyLLMWrapper"
```
Comment (skamanbo) [inline aibom/examples/.aibom.yaml:26]: Is there any precedence if multiple matches pop up?

File: aibom/src/aibom/categorizer.py [hunk 2]
```diff
@@ -266,13 +327,125 @@ def categorize_symbols(
...
+            # Merge user-provided metadata from custom catalog for this component
+            for custom_id, meta in custom_metadata_by_id.items():
+                if db_symbol_name.endswith(custom_id) or db_symbol_name == custom_id:
+                    component_details.update(meta)
+                    break
+
             _register_component_name(component_lookup_by_name, component_details["name"], component_details)
 
+    # ── Inline annotation pass: classes and functions tagged with # aibom: ──
+    for result in analysis_results:
```
Comment (skamanbo) [inline aibom/src/aibom/categorizer.py:339]: A symbol explicitly excluded in .aibom.yaml can still appear if it’s annotated or matches a base‑class rule?

File: aibom/examples/.aibom.yaml (hunk 1 above)
Comment (vineethsai7) [reply aibom/examples/.aibom.yaml:26]: duckdb > custom I think, but good point. Test case needs to be added.

File: aibom/src/aibom/categorizer.py (hunk 2 above)
Comment (vineethsai7) [reply aibom/src/aibom/categorizer.py:339]: I think exclusions override everything, but I would double checkl

File: aibom/examples/.aibom.yaml (hunk 1 above)
Comment (vineethsai7) [reply aibom/examples/.aibom.yaml:26]: Precedence is now defined and tested. Detection order:

1. **Inline annotations** (`# aibom: concept=...`) -- highest priority
2. **DuckDB catalog** (built-in knowledge base)
3. **Custom components** (`.aibom.yaml` entries)
4. **Base-class rules** -- lowest priority

If the same symbol at the same location is detected by multiple sources, only the highest-priority detection is kept. This is enforced via a `detected_locations` set in `categorizer.py` and covered by 6 new tests in `TestPrecedenceAndExcludes`.

File: aibom/src/aibom/categorizer.py (hunk 2 above)
Comment (vineethsai7) [reply aibom/src/aibom/categorizer.py:339]: Fixed. Excludes now override all detection sources -- inline annotations, DuckDB catalog, and base-class rules. Implementation:

- `_is_excluded()` helper checks each symbol before it enters any detection pass
- A final safety-net sweep runs before relationship derivation to catch any edge cases
- Three dedicated tests cover this: `test_excludes_filter_inline_annotations`, `test_excludes_filter_base_class_detections`, and the existing `test_excludes_filter_results` in `test_catalog_db.py`

**Comment by vineethsai7** [comment]
Superseded by comprehensive PR that consolidates all catalog improvements, bug fixes, and new framework coverage into a single changeset.
